File: backend/src/services/clientService.ts
import { Client } from '@prisma/client';
import { ApiError } from '../utils/errors';
import prisma from '../lib/prisma';

export class ClientService {
  async createClient(
    userId: string,
    data: {
      firstName: string;
      lastName: string;
      phone?: string;
      email?: string;
      address?: string;
    },
  ): Promise<Client> {
    return prisma.client.create({
      data: {
        userId,
        firstName: data.firstName,
        lastName: data.lastName,
        phone: data.phone,
        email: data.email,
        address: data.address,
      },
    });
  }

  async getClientById(clientId: string, userId: string): Promise<Client | null> {
    return prisma.client.findFirst({
      where: {
        id: clientId,
        userId,
      },
    });
  }

  async listClients(userId: string, options?: { skip?: number; take?: number }): Promise<Client[]> {
    return prisma.client.findMany({
      where: {
        userId,
        isActive: true,
      },
      skip: options?.skip,
      take: options?.take,
      orderBy: { createdAt: 'desc' },
    });
  }

  async updateClient(
    clientId: string,
    userId: string,
    data: Partial<Omit<Client, 'id' | 'userId' | 'createdAt' | 'updatedAt' | 'deletedAt'>>,
  ): Promise<Client> {
    const client = await this.getClientById(clientId, userId);

    if (!client) {
      throw new ApiError(404, 'Client not found');
    }

    return prisma.client.update({
      where: { id: clientId },
      data,
    });
  }

  async deleteClient(clientId: string, userId: string): Promise<Client> {
    const client = await this.getClientById(clientId, userId);

    if (!client) {
      throw new ApiError(404, 'Client not found');
    }

    return prisma.client.update({
      where: { id: clientId },
      data: {
        isActive: false,
        deletedAt: new Date(),
      },
    });
  }

  async getClientWithTransactions(clientId: string, userId: string) {
    const client = await prisma.client.findFirst({
      where: {
        id: clientId,
        userId,
      },
      include: {
        transactions: {
          where: { deletedAt: null },
          orderBy: { createdAt: 'desc' },
        },
      },
    });

    if (!client) {
      throw new ApiError(404, 'Client not found');
    }

    return client;
  }

  async searchClients(userId: string, query: string): Promise<Client[]> {
    return prisma.client.findMany({
      where: {
        userId,
        isActive: true,
        OR: [
          { firstName: { contains: query } },
          { lastName: { contains: query } },
          { phone: { contains: query } },
          { email: { contains: query } },
        ],
      },
      take: 10,
    });
  }

  async getClientStats(clientId: string, userId: string) {
    const client = await this.getClientById(clientId, userId);

    if (!client) {
      throw new ApiError(404, 'Client not found');
    }

    return {
      totalDebt: client.totalDebt,
      totalCredit: client.totalCredit,
      totalPayment: client.totalPayment,
    };
  }

  async setClientActiveStatus(clientId: string, userId: string, isActive: boolean): Promise<Client> {
    const client = await this.getClientById(clientId, userId);

    if (!client) {
      throw new ApiError(404, 'Client not found');
    }

    return prisma.client.update({
      where: { id: clientId },
      data: { isActive },
    });
  }
}

export default new ClientService();

File: front/lib/services/client_service.dart
import 'dart:convert';
import 'dart:async';
import 'package:http/http.dart' as http;
import '../models/client_model.dart';
import '../constants/app_constants.dart';
import 'api_client.dart';

class ClientService {
  static const String endpoint = '$baseUrl/clients';
  final ApiClient apiClient;
  final http.Client httpClient;

  ClientService({required this.apiClient, http.Client? httpClient})
    : httpClient = httpClient ?? http.Client();

  ApiException _mapException(Object error) {
    if (error is ApiException) {
      return error;
    }

    if (error is TimeoutException) {
      return ApiException(
        message: 'Request timeout. Please try again.',
        statusCode: 408,
      );
    }

    if (error is http.ClientException) {
      return ApiException(
        message: 'Network error. Please check your internet connection.',
        statusCode: 0,
      );
    }

    return ApiException(message: error.toString(), statusCode: 0);
  }

  Future<List<Client>> getClients({int skip = 0, int take = 10}) async {
    try {
      final response = await httpClient
          .get(
            Uri.parse('$endpoint?skip=$skip&take=$take'),
            headers: {'Authorization': 'Bearer ${apiClient.token}'},
          )
          .timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return (data['clients'] as List)
            .map((e) => Client.fromJson(e))
            .toList();
      } else {
        throw ApiException(
          message:
              jsonDecode(response.body)['error'] ?? 'Failed to load clients',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      throw _mapException(e);
    }
  }

  Future<Client> getClient(String clientId) async {
    try {
      final response = await httpClient
          .get(
            Uri.parse('$endpoint/$clientId'),
            headers: {'Authorization': 'Bearer ${apiClient.token}'},
          )
          .timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return Client.fromJson(data['client']);
      } else {
        throw ApiException(
          message: jsonDecode(response.body)['error'] ?? 'Client not found',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      throw _mapException(e);
    }
  }

  Future<Client> createClient({
    required String firstName,
    required String lastName,
    String? phone,
    String? email,
    String? address,
  }) async {
    try {
      final response = await httpClient
          .post(
            Uri.parse(endpoint),
            headers: {
              'Content-Type': 'application/json',
              'Authorization': 'Bearer ${apiClient.token}',
            },
            body: jsonEncode({
              'firstName': firstName,
              'lastName': lastName,
              'phone': phone,
              'email': email,
              'address': address,
            }),
          )
          .timeout(const Duration(seconds: 30));

      if (response.statusCode == 201) {
        final data = jsonDecode(response.body);
        return Client.fromJson(data['client']);
      } else {
        throw ApiException(
          message:
              jsonDecode(response.body)['error'] ?? 'Failed to create client',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      throw _mapException(e);
    }
  }

  Future<Client> updateClient(
    String clientId, {
    required String firstName,
    required String lastName,
    String? phone,
    String? email,
    String? address,
  }) async {
    try {
      final response = await httpClient
          .put(
            Uri.parse('$endpoint/$clientId'),
            headers: {
              'Content-Type': 'application/json',
              'Authorization': 'Bearer ${apiClient.token}',
            },
            body: jsonEncode({
              'firstName': firstName,
              'lastName': lastName,
              'phone': phone,
              'email': email,
              'address': address,
            }),
          )
          .timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return Client.fromJson(data['client']);
      } else {
        throw ApiException(
          message:
              jsonDecode(response.body)['error'] ?? 'Failed to update client',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      throw _mapException(e);
    }
  }

  Future<void> deleteClient(String clientId) async {
    try {
      final response = await httpClient
          .delete(
            Uri.parse('$endpoint/$clientId'),
            headers: {'Authorization': 'Bearer ${apiClient.token}'},
          )
          .timeout(const Duration(seconds: 30));

      if (response.statusCode != 200) {
        throw ApiException(
          message:
              jsonDecode(response.body)['error'] ?? 'Failed to delete client',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      throw _mapException(e);
    }
  }

  Future<List<Client>> searchClients(String query) async {
    try {
      final response = await httpClient
          .get(
            Uri.parse('$endpoint/search?q=$query'),
            headers: {'Authorization': 'Bearer ${apiClient.token}'},
          )
          .timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return (data['clients'] as List)
            .map((e) => Client.fromJson(e))
            .toList();
      } else {
        throw ApiException(
          message: jsonDecode(response.body)['error'] ?? 'Search failed',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      throw _mapException(e);
    }
  }

  Future<void> setClientStatus(String id, bool isActive) async {
    try {
      final response = await httpClient
          .patch(
            Uri.parse('$endpoint/$id/status'),
            headers: {
              'Content-Type': 'application/json',
              'Authorization': 'Bearer ${apiClient.token}',
            },
            body: jsonEncode({'isActive': isActive}),
          )
          .timeout(const Duration(seconds: 30));

      if (response.statusCode != 200) {
        throw ApiException(
          message:
              jsonDecode(response.body)['error'] ??
              'Failed to update client status',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      throw _mapException(e);
    }
  }
}
